File: cpp/cpp_100-999/591.cc
// C++
// leetcode 591
// https://leetcode-cn.com/problems/tag-validator/

class Solution {
public:
	bool is_big_letter(char c) {
		if(c >= 'A' && c <= 'Z')
			return true;
		return false;
	}

    bool isValid(string code) {
    	stack<string> tag_stk;
    	int n = code.size();
    	for(int i = 0; i < n; ) {
    		// tag_stk只能初始状态是空的，其他时候不行
    		if(tag_stk.empty() && (code[i] != '<' || i != 0)) {
    			return false;
    		}
    		// 找到第一个 <
    		while(i < n && code[i] != '<')
    			i++;
    		int j = i + 1;
    		// CDATA check
    		if(j + 7 < n && code.substr(j, 8) == "![CDATA[") {
    			j += 8;
    			bool have_an_end = false;
    			while(j < n) {
    				if(code[j] == ']') {
    					if(j + 2 >= n)
    						return false;
    					if(code[j+1] == ']' && code[j+2] == '>') {
    						// this CDATA part terminates
    						j += 2;
    						have_an_end = true;
    						break;
    					}
    				}
    				j++;
    			}
    			if(!have_an_end)
    				return false;
    			i = j+1;
    			continue;
    		}
    		// <> 的 check
    		if(j == n)
    			return false;
    		int left_cnt = 1;
    		bool out;
    		if(code[j] == '/') {
    			out = true;
    			i++;
    			j++;
    		} else {
    			out = false;
    		}

  	  		while(j < n && left_cnt != 0) {
  	  			char c = code[j];
  	  			if(c == '<') {
  	  				left_cnt++;
  	  			} else if(c == '>') {
  	  				left_cnt--;
  	  			}else if(!is_big_letter(c)) {
  	  				return false;
  	  			}
  	  			j++;
  	  		}
  	  		
  	  		if(left_cnt != 0)
  	  			return false;
  	  		int tmp_size = j-i-2;
            // cout<<tmp_size<<endl;
  	  		if(tmp_size <1 || tmp_size > 9)
  	  			return false;
  	  		string tmp = code.substr(i+1, tmp_size);
            // cout<<tmp<<endl;
  	  		if(out) {
  	  			if(tag_stk.empty() || tag_stk.top() != tmp)
  	  				return false;
  	  			tag_stk.pop();
  	  		} else {
  	  			tag_stk.push(tmp);
  	  		}
  	  		i = j;
    	}
    	return tag_stk.empty();
    }
};

File: Go/Go_100-999/591.go
// Go
// leetcode 591
// https://leetcode.cn/problems/tag-validator/

func isValid(code string) bool {
	n := len(code)
	tagStk := []string{}
	if code[0] != '<' {
		return false
	}
	for i := 0; i < n; i++ {
		if len(tagStk) == 0 && i != 0 {
			return false
		}

		// find first <
		for i < n && code[i] != '<' {
			i++
		}

		// CDATA check
		if i+8 < n && code[i:i+9] == "<![CDATA[" {
			i += 9
			for i+2 < n && code[i:i+3] != "]]>" {
				i++
			}
			if i+2 >= n {
				return false
			}
			i += 2
			continue
		}

		if i+1 < n && code[i+1] == '/' { // 结束标签
			i += 2
			j := i
			for j < n && code[j] != '>' {
				if !isBigLetter(code[j]) || j-i > 8 {
					return false
				}
				j++
			}
			if j == n || i == j {
				return false
			}
			tag := code[i:j]
			if len(tagStk) == 0 || tagStk[len(tagStk)-1] != tag {
				return false
			}
			tagStk = tagStk[:len(tagStk)-1]
			i = j
			continue
		}

		// 起始标签
		i++
		j := i
		for j < n && code[j] != '>' {
			if !isBigLetter(code[j]) || j-i > 8 {
				return false
			}
			j++
		}
		if j == n || i == j {
			return false
		}
		tag := code[i:j]
		tagStk = append(tagStk, tag)
		i = j
	}
	return len(tagStk) == 0
}

func isBigLetter(b byte) bool {
	return b <= 'Z' && b >= 'A'
}